obs/stats.go

```diff
@@ -0,0 +1,343 @@
+package obs
+
+import (
+	"context"
+	"strings"
+
+	logging "github.com/ipfs/go-log/v2"
+	rcmgr "github.com/libp2p/go-libp2p-resource-manager"
+	"go.opencensus.io/stats"
+	"go.opencensus.io/stats/view"
+	"go.opencensus.io/tag"
+)
+
+var log = logging.Logger("rcmgrObs")
+
+var (
+	systemOutboundConns = stats.Int64("system/outbound/conn", "Number of outbound Connections", stats.UnitDimensionless)
+	systemInboundConns  = stats.Int64("system/inbound/conn", "Number of inbound Connections", stats.UnitDimensionless)
+
+	conns = stats.Int64("connections", "Number of Connections", stats.UnitDimensionless)
+
+	peerConns         = stats.Int64("peer/connections", "Number of connections this peer has", stats.UnitDimensionless)
+	peerConnsNegative = stats.Int64("peer/connections_negative", "Number of connections this peer had", stats.UnitDimensionless)
+
+	streams = stats.Int64("streams", "Number of Streams", stats.UnitDimensionless)
+
+	peerStreams         = stats.Int64("peer/streams", "Number of streams this peer has", stats.UnitDimensionless)
+	peerStreamsNegative = stats.Int64("peer/streams_negative", "Number of streams this peer had", stats.UnitDimensionless)
+
+	memory             = stats.Int64("memory", "Amount of memory reserved as reported to the Resource Manager", stats.UnitDimensionless)
+	peerMemory         = stats.Int64("peer/memory", "Amount of memory currently reseved for peer", stats.UnitDimensionless)
+	peerMemoryNegative = stats.Int64("peer/memory_negative", "Amount of memory previously reseved for peer", stats.UnitDimensionless)
+
+	connMemory         = stats.Int64("conn/memory", "Amount of memory currently reseved for the connection", stats.UnitDimensionless)
+	connMemoryNegative = stats.Int64("conn/memory_negative", "Amount of memory previously reseved for the connection", stats.UnitDimensionless)
+
+	fds = stats.Int64("fds", "Number of fds as reported to the Resource Manager", stats.UnitDimensionless)
+
+	blockedResources = stats.Int64("blocked_resources", "Number of resource requests blocked", stats.UnitDimensionless)
+)
+
+var (
+	LessThanEq, _ = tag.NewKey("le")
+	Direction, _  = tag.NewKey("dir")
+	Scope, _      = tag.NewKey("scope")
+	Service, _    = tag.NewKey("service")
+	Protocol, _   = tag.NewKey("protocol")
+	Resource, _   = tag.NewKey("resource")
+)
+
+var (
+	SystemOutboundConnsView = &view.View{Measure: systemOutboundConns, Aggregation: view.Sum()}
+	SystemInboundConnsView  = &view.View{Measure: systemInboundConns, Aggregation: view.Sum()}
+
+	ConnView = &view.View{Measure: conns, Aggregation: view.Sum(), TagKeys: []tag.Key{Direction, Scope}}
+
+	fibLikeDistribution = []float64{
+		1.1, 2.1, 3.1, 5.1, 8.1, 13.1, 21.1, 34.1, 55.1, 100.1, 200.1,
+	}
```

Comment:
Out of curiosity, why this distribution?

Reply:
fibonaccci with +0.1 because of a quirk in opencensus that does bound non-inclusive bound checking: https://github.com/census-instrumentation/opencensus-go/blob/v0.23.0/stats/view/aggregation_data.go#L195.

But the fib sequence itself is totally arbitrary. Happy to change this to something else if you have a better idea.

Reply:
No better ideas. The go-to distributions are _linear_ and _exponential_ in the Prometheus ecosystem:

https://github.com/prometheus/client_golang/blob/4ad265f1b4ee8d5c239b3f36081529bbe3a838db/prometheus/histogram.go#L75-L145